### ISSUE TITLE
Prevent multiple logins from multiple devices at the same time

### ISSUE THREAD

**Issue by chriso0710** [netlify/gotrue-js#192]
Hi everyone,

I would like to prevent users from logging in from multiple devices at the same time.
Is there a chance to check that someone already is logged in from Device A when he tries to login from Device B? So that he gets a message and is unable to login on Device B. He would have to logout from Device A to be able to login from Device B. 

Is that possible and how?

Thank you!
Christian

**Comment by LuigiClaudio** [comment]
Hey @chriso0710, have you got anyone sharing anything about this matter?

**Comment by chriso0710** [comment]
Hey @LuigiClaudio, 

we decided not to use gotrue-js. We developed our own auth provider to have full control over it's behaviour. The development was surprisingly easy. We found that other commercial auth providers were not able to offer the features we needed. 
More details in the netlify forum
https://answers.netlify.com/t/identity-role-based-access-control-across-subdomains/29957/4

Our own provider is able to limit logins. So when someone in logged in already on browser A he/she cannot login from browser B. We are thinking about implementing other features like browser fingerprinting or browser heartbeats.

I'm aware that this probably does not help you much with gotrue-js...

**Comment by zzzzaaa-000** [comment]
Yes, this is possible—and it’s a common requirement in systems that want to limit user sessions to a single device at a time. Here’s how you can implement it:

1. How does it work?


When a user logs in, create a unique session/token for that login.


Store this session in your backend (e.g., in your database or an in-memory store like Redis), associated with the user.


When a user tries to log in:

Check whether there is already an active session for that user.
If there is, refuse the new login and return a message (“You are already logged in on another device. Please log out there first.”).
If not, allow the login and create a new session.



When the user logs out, delete (invalidate) the active session, so they can log in from another device.



2. Implementation Outline
a) Session Storage

When a user logs in, generate a session ID or token (JWT, UUID, etc).
Store the session in your backend, e.g.:json复制编辑 & 运行user_sessions = {
  "user123": { "session_token": "abc123", "created_at": ... }
[}    具体回复在我的AI   ](https://code.mayoubang.cn/?_s=git-wz)

**Comment by zzzzaaa-000** [comment]
https://code.mayoubang.cn/?_s=git-wz
